RecoTauTag/RecoTau/plugins/PFRecoTauMassPlugin.cc

```diff
@@ -0,0 +1,82 @@
+/*
+ * =============================================================================
+ *       Filename:  PFRecoTauMassPlugin.cc
+ *
+ *    Description:  Set mass of taus reconstructed in 1prong0pi0 decay mode
+ *                  to charged pion mass
+ *
+ *        Created:  27/10/2015 10:30:00
+ *
+ *         Authors:  Christian Veelken (Tallinn)
+ *
+ * =============================================================================
+ */
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "RecoTauTag/RecoTau/interface/RecoTauBuilderPlugins.h"
+#include "DataFormats/TauReco/interface/PFTau.h"
+
+#include <TMath.h>
+
+namespace reco { namespace tau {
+
+class PFRecoTauMassPlugin : public RecoTauModifierPlugin
+{
+ public:
+
+  explicit PFRecoTauMassPlugin(const edm::ParameterSet&, edm::ConsumesCollector &&iC);
+  virtual ~PFRecoTauMassPlugin();
+  void operator()(PFTau&) const;
+  virtual void beginEvent();
+  virtual void endEvent();
+
+ private:
+  
+  int verbosity_;
+};
+
+  PFRecoTauMassPlugin::PFRecoTauMassPlugin(const edm::ParameterSet& cfg, edm::ConsumesCollector &&iC)
+    : RecoTauModifierPlugin(cfg, std::move(iC))
+{
+  verbosity_ = ( cfg.exists("verbosity") ) ?
+    cfg.getParameter<int>("verbosity") : 0;
+}
+
+PFRecoTauMassPlugin::~PFRecoTauMassPlugin()
+{}
+
+void PFRecoTauMassPlugin::beginEvent()
+{}
+
+void PFRecoTauMassPlugin::operator()(PFTau& tau) const
+{
+  if ( verbosity_ ) {
+    std::cout << "<PFRecoTauMassPlugin::operator()>:" << std::endl;
+    std::cout << "tau: Pt = " << tau.pt() << ", eta = " << tau.eta() << ", phi = " << tau.phi() << ", mass = " << tau.mass() << " (decayMode = " << tau.decayMode() << ")" << std::endl;
+  }
+
+  if ( tau.decayMode() == reco::PFTau::kOneProng0PiZero ) {
+    double tauEn = tau.energy();
+    const double chargedPionMass = 0.13957; // GeV
+    if ( tauEn < chargedPionMass ) tauEn = chargedPionMass;
+    double tauP_modified = TMath::Sqrt(tauEn*tauEn - chargedPionMass*chargedPionMass);
+    double tauPx_modified = TMath::Cos(tau.phi())*TMath::Sin(tau.theta())*tauP_modified;
```

Comment:
what's wrong with
```tauPx_modified = tau.px()*tauP_modified/ tau.p(); ```?
you are loosing in precision by making an extra turnaround through sin/cos(atan(x,y))